nifi-nar-bundles/nifi-druid-bundle/nifi-druid-controller-service/src/main/java/org/apache/nifi/controller/DruidTranquilityController.java

```diff
@@ -0,0 +1,416 @@
+	
+	static{
+		final List<PropertyDescriptor> props = new ArrayList<>();
+	    props.add(DATASOURCE);
+	    props.add(CONNECT_STRING);
+	    props.add(DRUID_INDEX_SERVICE_PATH);
+	    props.add(DRUID_DISCOVERY_PATH);
+	    props.add(DIMENSIONS_LIST);
+	    props.add(AGGREGATOR_JSON);
+	    props.add(SEGMENT_GRANULARITY);
+	    props.add(QUERY_GRANULARITY);
+	    props.add(WINDOW_PERIOD);
+	    props.add(TIMESTAMP_FIELD);
+	    
+	    properties = Collections.unmodifiableList(props);
+	}
+	
+	@Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+	@OnEnabled
+	public void onConfigured(final ConfigurationContext context) throws InitializationException{
+		getLogger().info("Starting Druid Tranquility Controller Service...");
+	   
+		final String dataSource = context.getProperty(DATASOURCE).getValue();
+		final String zkConnectString = context.getProperty(CONNECT_STRING).getValue();
+		final String indexService = context.getProperty(DRUID_INDEX_SERVICE_PATH).getValue();
+		final String discoveryPath = context.getProperty(DRUID_DISCOVERY_PATH).getValue();
+		final String timestampField  = context.getProperty(TIMESTAMP_FIELD).getValue();
+		final String segmentGranularity = context.getProperty(SEGMENT_GRANULARITY).getValue();
+		final String queryGranularity = context.getProperty(QUERY_GRANULARITY).getValue();
+		final String windowPeriod = context.getProperty(WINDOW_PERIOD).getValue();
+		final String aggregatorJSON = context.getProperty(AGGREGATOR_JSON).getValue();
+		final String dimensionsStringList = context.getProperty(DIMENSIONS_LIST).getValue();
+		
+		final List<String> dimensions = getDimensions(dimensionsStringList);
+	    final List<AggregatorFactory> aggregator = getAggregatorList(aggregatorJSON);
+	    
+		final Timestamper<Map<String, Object>> timestamper = new Timestamper<Map<String, Object>>(){
+			private static final long serialVersionUID = 1L;
+
+			@Override
+			public DateTime timestamp(Map<String, Object> theMap){
+				return new DateTime(theMap.get(timestampField));
+			}
+		};
+		
+		Iterator<AggregatorFactory> aggIterator = aggregator.iterator();
+		AggregatorFactory currFactory;
+		getLogger().debug("Number of Aggregations Defined: " + aggregator.size());
+		while(aggIterator.hasNext()){
+			currFactory = aggIterator.next();
+			getLogger().debug("Verifying Aggregator Definition");
+			getLogger().debug("Aggregator Name: " + currFactory.getName());
+			getLogger().debug("Aggregator Type: " + currFactory.getTypeName());
+			getLogger().debug("Aggregator Req Fields: " + currFactory.requiredFields());			
+		}
+		// Tranquility uses ZooKeeper (through Curator) for coordination.
+		final CuratorFramework curator = CuratorFrameworkFactory
+		                .builder()
+		                .connectString(zkConnectString)
+		                .retryPolicy(new ExponentialBackoffRetry(1000, 20, 30000))
+		                .build();
+		curator.start();
```

Comment:
we'd create a connect/session to zk with this right?  Do we ever close it?

Reply:
I believe Tranquility needs this connection open for the life of the Service. It constantly needs to see the meta data there (servedSegments, coordinator, listeners, loadQueue, discovery, indexer, announcements, segments). The Storm Tranquility implementation is the same.
https://github.com/apache/storm/blob/master/external/storm-druid/src/test/java/org/apache/storm/druid/SampleDruidBeamFactoryImpl.java#L83